state/immutable-trie/snapshot.go

```diff
@@ -0,0 +1,78 @@
+package itrie
+
+import (
+	"github.com/0xPolygon/polygon-edge/crypto"
+	"github.com/0xPolygon/polygon-edge/state"
+	"github.com/0xPolygon/polygon-edge/types"
+	"github.com/umbracle/fastrlp"
+)
+
+type Snapshot struct {
+	state *State
+	trie  *Trie
+}
+
+var emptyStateHash = types.StringToHash("0x56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")
+
+func (s *Snapshot) GetStorage(addr types.Address, root types.Hash, rawkey types.Hash) types.Hash {
+	var (
+		err  error
+		trie *Trie
+	)
+
+	if root == emptyStateHash {
+		trie = s.state.newTrie()
+	} else {
+		trie, err = s.state.newTrieAt(root)
+		if err != nil {
+			return types.Hash{}
```

Comment:
Is it correct?

Reply:
Kind of, yes. The EVM does not have a way to signal that there was an error (and it has to stop) due to an external condition (like storage failure). If it fails, the storage entry returned will be wrong and the computation will be wrong so either the block created will not be valid (block creator) or it will not be able to validate a block.